strategy.py

```diff
@@ -1,52 +1,68 @@
 import logging
 from abc import ABC, abstractmethod
 from collections import defaultdict
-from datetime import timedelta
+from datetime import datetime, timedelta
 from typing import List, Union
 
 import numpy as np
 import pandas as pd
 import simpy
 from leaf.infrastructure import Node
 
+from vessim import TimeSeriesApi
+
 logger = logging.getLogger(__name__)
 
 
 class Strategy(ABC):
     replicated_bits = 0
 
-    def __init__(self, node: Node):
+    def __init__(self, node: Node, ci: pd.Series, interval: int) -> None:
         self.node = node
+        self.api = TimeSeriesApi(actual=ci)
+        self.interval = interval
 
     @abstractmethod
     def run(self, env: simpy.Environment, duration: int):
         pass
 
+    def _start_time(self, env: simpy.Environment) -> pd.Timestamp:
+        return (
+            pd.Timestamp(self.api._actual.index[0]) + 
+            pd.Timedelta(minutes=env.now)
+        )
+
+    def _forecast(self, start_time: datetime, end_time: datetime) -> pd.Series:
+        return pd.concat([
+            pd.Series(self.api.actual(start_time), index=[start_time]),
+            self.api.forecast(start_time, end_time, frequency=f"{self.interval}min")
+        ])
 
 class BidirectionalStrategy(Strategy):
-    """A strategy for scheduled workloads in the future, i.e. workloads that can be shifted in both "directions"."""
+    """A strategy for scheduled workloads in the future. Workloads that can be shifted in both "directions"."""
 
-    def __init__(self, node: Node, ci: List[float], window_in_minutes: float, interval: int):
-        super().__init__(node)
-        self.ci = ci
+    def __init__(self, node: Node, ci: pd.DataFrame, window_in_minutes: float, interval: int):
+        super().__init__(node, ci, interval)
         self.window_in_minutes = window_in_minutes
-        self.interval = interval
-        self.forecast_steps = int(window_in_minutes / self.interval)
         self.hist = np.zeros(48)  # 48 steps per day
 
     def run(self, env: simpy.Environment, duration: int):
         assert duration == self.interval
-        timeline_idx = env.now // self.interval
-        forecast = self.ci[timeline_idx:timeline_idx + self.forecast_steps * 2 + 1]
+        start_time = self._start_time(env)
+        end_time = start_time + pd.Timedelta(minutes=self.window_in_minutes * 2)
+        forecast = self._forecast(start_time, end_time)
 
-        wait_until = np.argmin(forecast)
-        timestamp = forecast.index[wait_until]
+        wait_index = np.argmin(forecast)
+        timestamp = forecast.index[wait_index]
         minute = 0 if timestamp.minute == 0 else 1
         self.hist[timestamp.hour * 2 + minute] += 1
 
-        logger.debug(f"{env.now}: Wait {wait_until * self.interval}h (current: {self.ci[timeline_idx]}, "
-                     f"then {self.ci[timeline_idx + wait_until]}).")
-        yield env.timeout(wait_until * self.interval)
+        wait_minutes = (timestamp - start_time).total_seconds() / 60
```

Comment:
hours, no?

Reply:
`wait_until` should be an index of the `forecast` dataframe which has data in `self.interval` (30 in this case) minute intervals. So `wait_until * self.interval` should make up minutes. If I didn't have bad luck with my thinking there I should either change the `h` string in the log message of divide by 60. What do you think?

Reply:
both is fine, let's stay with "h" then to be consistent with the old code :)